File: src/main/java/com/sedmelluq/discord/lavaplayer/player/FunctionalResultHandler.java
package com.sedmelluq.discord.lavaplayer.player;

import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;

import java.util.function.Consumer;

/**
 * Helper class for creating an audio result handler using only methods that can be passed as lambdas.
 */
public class FunctionalResultHandler implements AudioLoadResultHandler {
  private final Consumer<AudioTrack> trackConsumer;
  private final Consumer<AudioPlaylist> playlistConsumer;
  private final Runnable emptyResultHandler;
  private final Consumer<FriendlyException> exceptionConsumer;

  /**
   * Refer to {@link AudioLoadResultHandler} methods for details on when each method is called.
   *
   * @param trackConsumer Consumer for single track result
   * @param playlistConsumer Consumer for playlist result
   * @param emptyResultHandler Empty result handler
   * @param exceptionConsumer Consumer for an exception when loading the item fails
   */
  public FunctionalResultHandler(Consumer<AudioTrack> trackConsumer, Consumer<AudioPlaylist> playlistConsumer,
                                 Runnable emptyResultHandler, Consumer<FriendlyException> exceptionConsumer) {

    this.trackConsumer = trackConsumer;
    this.playlistConsumer = playlistConsumer;
    this.emptyResultHandler = emptyResultHandler;
    this.exceptionConsumer = exceptionConsumer;
  }

  @Override
  public void trackLoaded(AudioTrack track) {
    if (trackConsumer != null) {
      trackConsumer.accept(track);
    }
  }

  @Override
  public void playlistLoaded(AudioPlaylist playlist) {
    if (playlistConsumer != null) {
      playlistConsumer.accept(playlist);
    }
  }

  @Override
  public void noMatches() {
    if (emptyResultHandler != null) {
      emptyResultHandler.run();
    }
  }

  @Override
  public void loadFailed(FriendlyException exception) {
    if (exceptionConsumer != null) {
      exceptionConsumer.accept(exception);
    }
  }
}

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/player/FunctionalResultHandler.kt
package com.sedmelluq.discord.lavaplayer.player;

import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackCollection;

import java.util.function.Consumer;

/**
 * Helper class for creating an audio result handler using only methods that can be passed as lambdas.
 *
 * @param trackConsumer      Consumer for single track result
 * @param playlistConsumer   Consumer for playlist result
 * @param emptyResultHandler Empty result handler
 * @param exceptionConsumer  Consumer for an exception when loading the item fails
 */
class FunctionalResultHandler(
    val trackConsumer: Consumer<AudioTrack>?,
    val emptyResultHandler: Runnable?,
    val exceptionConsumer: Consumer<FriendlyException>?,
    val collectionConsumer: Consumer<AudioTrackCollection>? = null,
) : AudioLoadResultHandler {

    override fun trackLoaded(track: AudioTrack) {
        trackConsumer?.accept(track)
    }

    override fun collectionLoaded(collection: AudioTrackCollection) {
        collectionConsumer?.accept(collection)
    }

    override fun noMatches() {
        emptyResultHandler?.run()
    }

    override fun loadFailed(exception: FriendlyException) {
        exceptionConsumer?.accept(exception)
    }
}
